File: src/inventory/inventory.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { SupabaseService } from 'src/supabase/supabase.service';
import { PrismaService } from 'src/prisma/prisma.service';
import { CreateInventoryItemDto } from './dto/create-inventory-item.dto';
import { UpdateInventoryItemDto } from './dto/update-inventory-item.dto';

@Injectable()
export class InventoryService {
  constructor(
    private readonly supabaseService: SupabaseService,
    private readonly prisma: PrismaService,
  ) {}

  async create(userId: string, dto: CreateInventoryItemDto) {
    return this.prisma.inventory_items.create({
      data: {
        user_id: userId,
        item_name: dto.itemName,
        quantity: dto.quantity,
        unit: dto.unit,
        average_cost: dto.averageCost,
      },
    });
  }

  async findAll(userId: string) {
    return this.prisma.inventory_items.findMany({
      where: {
        user_id: userId,
      },
      orderBy: {
        updated_at: 'desc',
      },
    });
  }

  async findOne(id: bigint, userId: string) {
    return this.assertOwnership(id, userId);
  }

  async update(id: bigint, userId: string, dto: UpdateInventoryItemDto) {
    await this.assertOwnership(id, userId);

    return this.prisma.inventory_items.update({
      where: {
        id,
      },
      data: {
        item_name: dto.itemName,
        quantity: dto.quantity,
        unit: dto.unit,
        average_cost: dto.averageCost,
        updated_at: new Date(),
      },
    });
  }

  async remove(id: bigint, userId: string) {
    await this.assertOwnership(id, userId);

    return this.prisma.inventory_items.delete({
      where: {
        id,
      },
    });
  }

  async adjustStock(id: bigint, userId: string, quantity: number) {
    const item = await this.assertOwnership(id, userId);

    const newQuantity = Number(item.quantity ?? 0) + quantity;

    if (newQuantity < 0) {
      throw new BadRequestException(
        'Adjustment would result in negative stock',
      );
    }

    return this.prisma.inventory_items.update({
      where: {
        id,
      },
      data: {
        quantity: newQuantity,
        updated_at: new Date(),
      },
    });
  }

  private async assertOwnership(id: bigint, userId: string) {
    const item = await this.prisma.inventory_items.findUnique({
      where: {
        id,
      },
    });

    if (!item) {
      throw new NotFoundException('Inventory item not found');
    }

    if (item.user_id !== userId) {
      throw new ForbiddenException(
        'You do not have access to this inventory item',
      );
    }

    return item;
  }
}

File: src/services/inventory_service.py
from fastapi import HTTPException

from src.schemas.inventory import AdjustStockRequest, CreateInventoryItemRequest, UpdateInventoryItemRequest
from src.services.supabase_client import get_supabase_client


def create_item(user_id: str, body: CreateInventoryItemRequest) -> dict:
    client = get_supabase_client()
    result = (
        client.table("inventory_items")
        .insert(
            {
                "user_id": user_id,
                "item_name": body.item_name,
                "quantity": body.quantity,
                "unit": body.unit,
                "average_cost": body.average_cost,
            }
        )
        .execute()
    )
    return result.data[0]


def find_all_items(user_id: str) -> list[dict]:
    client = get_supabase_client()
    result = (
        client.table("inventory_items")
        .select("*")
        .eq("user_id", user_id)
        .order("updated_at", desc=True)
        .execute()
    )
    return result.data


def find_item(item_id: int) -> dict:
    client = get_supabase_client()
    result = client.table("inventory_items").select("*").eq("id", item_id).execute()
    if not result.data:
        raise HTTPException(status_code=404, detail="Inventory item not found")
    return result.data[0]


def update_item(item_id: int, body: UpdateInventoryItemRequest) -> dict:
    client = get_supabase_client()
    updates = {k: v for k, v in body.model_dump().items() if v is not None}
    result = client.table("inventory_items").update(updates).eq("id", item_id).execute()
    if not result.data:
        raise HTTPException(status_code=404, detail="Inventory item not found")
    return result.data[0]


def delete_item(item_id: int) -> dict:
    client = get_supabase_client()
    result = client.table("inventory_items").delete().eq("id", item_id).execute()
    if not result.data:
        raise HTTPException(status_code=404, detail="Inventory item not found")
    return result.data[0]


def adjust_stock(item_id: int, body: AdjustStockRequest) -> dict:
    item = find_item(item_id)
    new_quantity = float(item["quantity"] or 0) + body.quantity

    client = get_supabase_client()
    result = client.table("inventory_items").update({"quantity": new_quantity}).eq("id", item_id).execute()
    return result.data[0]
